### PR TITLE
[#IR-306] As a dev, Add functionality in back-end so that we can validate Email. 

### PR DESCRIPTION
**Pull request by chahat-arora**
Also, generate and send OTP to the email address and also added a model to store OTP.

### CODE REVIEW

File: apps/user/models/otp.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+func RegisterOTPToAdmin(Admin *admin.Admin, config admin.Config) {
+	otp := Admin.AddResource(&OTP{}, &config)
+
+	otp.NewAttrs("-Code")
+	otp.EditAttrs("-Code")
+}
+
+// BeforeCreate ...
+func (otp *OTP) BeforeCreate(scope *gorm.Scope) error {
+	//generate a hexa decimal value for "code" using time.Now().Unix().
```
Comment (himanshu-garg-165) [inline apps/user/models/otp.go]:
```suggestion
	//generate a hexadecimal value for "code" using time.Now().UnixNano().
```

File: apps/user/models/otp.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	err := scope.SetColumn("Code", fmt.Sprintf("%X", time.Now().UnixNano()))
+	if err != nil {
+		return err
+	}
+	expiryAt := time.Unix(time.Now().Unix()+constants.OTPExpiryTime, 0)
```
Comment (himanshu-garg-165) [inline apps/user/models/otp.go:36]:
```suggestion
	expiryAt := time.Unix(time.Now().Unix() + constants.OTPExpiryTime, 0)
```

File: apps/user/serializers/user.go [hunk 3]
```diff
@@ -40,3 +40,9 @@ type UserLogin struct {
 	Email    string `json:"Email"`
 	Password string `json:"Password"`
 }
+
+// Identify ...
+type Identify struct {
+	Email   string `json:"email"`
+	SendOTP bool   `json:"sendOTP"`
```
Comment (himanshu-garg-165) [inline apps/user/serializers/user.go]:
```suggestion
	CreateOTP bool   `json:"createOTP"`
```

File: apps/user/services/authentication.go [hunk 4]
```diff
@@ -89,6 +93,102 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
+		}
+		return http.StatusInternalServerError, err
+	}
+	// when we don't need OTP to the email.
+	if !identifyData.SendOTP {
+		return http.StatusOK, nil
+	}
+	var otp userModels.OTP
+	err = gormDB.Model(&userModels.OTP{}).Where("user_id = ?", userData.ID).Scan(&otp).Error
+	if err != nil && err != gorm.ErrRecordNotFound {
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: We should use the `isRecordNotFoundError()` method here.

File: apps/user/services/authentication.go [hunk 5]
```diff
@@ -89,6 +93,102 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
+		return http.StatusInternalServerError, err
+	}
+	if err != gorm.ErrRecordNotFound {
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Why are we checking the same condition again?

File: apps/user/services/authentication.go [hunk 6]
```diff
@@ -89,6 +93,102 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	return http.StatusOK, nil
+}
+
+func sendOTPAtEmail(email string, code string, firstName string, lastName string) error {
+	message, _ := parseTemplate("apps/user/views/mail.html", map[string]interface{}{"firstName": firstName, "lastName": lastName, "code": code})
+	subject := "Subject: " + "One Time Password" + "\n"
+	from := "From: iReflect<no-reply@ireflect.com>\n"
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Why are we not sending the  `To` header?

File: apps/user/views/mail.html [hunk 7]
```diff
@@ -0,0 +1,52 @@
+<html xmlns="http://www.w3.org/1999/xhtml">
```
Comment (himanshu-garg-165) [inline apps/user/views/mail.html]: Why is `xmlns` needed?

File: apps/user/views/mail.html [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        <td style="padding:10px;">
+          <br>
+            Hi <b> {{.firstName}} {{.lastName}}</b>, <br/>
+            <b>{{.code}}</b> is your one time password (OTP). Please enter the OTP to proceed.<br/><br/>
+            Thank You,<br/>
+            Team iReflect.
+          </p>
+        </td>
+      </tr>
+      <tr class="footer">
```
Comment (himanshu-garg-165) [inline apps/user/views/mail.html]: Please remove it as discussed.

File: constants/auth.go [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package constants
+
+// OTPExpiryTime is OTP expiry time in seconds...
+const OTPExpiryTime = 300
+
+// OTPReCreationTime ...
+const OTPReCreationTime = 120
+
+// EmailUsername ...
+const EmailUsername = "sandipagarwal"
```
Comment (himanshu-garg-165) [inline constants/auth.go]: Please move them (username and password) to environment variables.

File: db/base/models/otp.go [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package models
+
+import "time"
+
+// OTP ...
+type OTP struct {
+	Code       string `gorm:"type:varchar(16);not null"`
+	ExpiryAt time.Time
```
Comment (himanshu-garg-165) [inline db/base/models/otp.go:8]: Please format the file properly.

File: apps/user/services/authentication.go [hunk 11]
```diff
@@ -89,6 +93,102 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
+	err = sendOTPAtEmail(identifyData.Email, newOTP.Code, userData.FirstName, userData.LastName)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	return http.StatusOK, nil
+}
+
+func sendOTPAtEmail(email string, code string, firstName string, lastName string) error {
+	message, _ := parseTemplate("apps/user/views/mail.html", map[string]interface{}{"firstName": firstName, "lastName": lastName, "code": code})
+	subject := "Subject: " + "One Time Password" + "\n"
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Pick all the constants from the constants file.

**Review by himanshu-garg-165 — CHANGES_REQUESTED** [review]

File: apps/user/services/authentication.go (hunk 11 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: done.

File: db/base/models/otp.go (hunk 10 above)
Comment (chahat-arora) [reply db/base/models/otp.go:8]: done.

File: constants/auth.go (hunk 9 above)
Comment (chahat-arora) [reply constants/auth.go]: done.

File: apps/user/views/mail.html (hunk 8 above)
Comment (chahat-arora) [reply apps/user/views/mail.html]: removed

File: apps/user/views/mail.html (hunk 7 above)
Comment (chahat-arora) [reply apps/user/views/mail.html]: it's optional. removed.

File: apps/user/services/authentication.go (hunk 6 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: done.

File: apps/user/services/authentication.go (hunk 5 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: check if OTP exists.

File: apps/user/services/authentication.go (hunk 4 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: done.

File: apps/user/serializers/user.go (hunk 3 above)
Comment (chahat-arora) [reply apps/user/serializers/user.go]: This boolean is to decide that we send OTP via email or not. In any case, OTP is created.

File: apps/user/models/otp.go (hunk 2 above)
Comment (chahat-arora) [reply apps/user/models/otp.go:36]: It is wrong according to go formatting scheme.

File: apps/user/models/otp.go (hunk 1 above)
Comment (chahat-arora) [reply apps/user/models/otp.go]: done.

File: apps/user/services/authentication.go [hunk 12]
```diff
@@ -90,6 +95,103 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
 
+// Identify ...
+func (service AuthenticationService) Identify(c *gin.Context) (
+	status int,
+	err error) {
+
+	var identifyData userSerializers.Identify
+	err = c.BindJSON(&identifyData)
+	if err != nil {
+		return http.StatusInternalServerError, err
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Return a 400 here.

File: apps/user/services/authentication.go [hunk 13]
```diff
@@ -90,6 +95,103 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
+	if !identifyData.SendOTP {
+		return http.StatusOK, nil
+	}
+	var otp userModels.OTP
+	err = gormDB.Model(&userModels.OTP{}).Where("user_id = ?", userData.ID).Scan(&otp).Error
+	if err != nil && !gorm.IsRecordNotFoundError(err) {
+		return http.StatusInternalServerError, err
+	}
+	// if OTP exists then check its validity.
+	if !gorm.IsRecordNotFoundError(err) {
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Doesn't the above condition `if err != nil && !gorm.IsRecordNotFoundError(err)` confirm that an OTP exists for the user?

File: apps/user/services/authentication.go [hunk 14]
```diff
@@ -90,6 +95,103 @@ func (service AuthenticationService) BasicLogin(c *gin.Context) (
...
+		emailConfig.Password,
+		emailConfig.Host,
+	)
+
+	// Connect to the server, authenticate, set the sender and recipient,
+	// and send the email all in one step.
+	err := smtp.SendMail(
+		fmt.Sprintf("%s:%s", emailConfig.Host, emailConfig.Port),
+		auth,
+		"no-reply@ireflect.com",
```
Comment (himanshu-garg-165) [inline apps/user/services/authentication.go]: Please make it a constant.

File: constants/constants.go [hunk 15]
```diff
@@ -87,3 +87,14 @@ var ActionTypeMap = map[ActionType]string{
...
 	InvalidEmailOrPassword = "Invalid email or password"
 )
+
+// <----------- constants for email --------------->
+
+// OTPEmailSubject ...
+const OTPEmailSubject = "Subject: One Time Password\n"
+
+// OTPEmailFrom ...
+const OTPEmailFrom = "From: iReflect<no-reply@ireflect.com>\n"
```
Comment (himanshu-garg-165) [inline constants/constants.go]: This should be same for the complete iReflct application.
```suggestion
const EmailFrom = "From: iReflect<no-reply@ireflect.com>\n"
```

File: constants/constants.go [hunk 16]
```diff
@@ -87,3 +87,14 @@ var ActionTypeMap = map[ActionType]string{
...
+
+// OTPEmailMIME ...
+const OTPEmailMIME = "MIME-version: 1.0;\nContent-Type: text/html; charset=\"UTF-8\";\n\n"
```
Comment (himanshu-garg-165) [inline constants/constants.go]: Similarly for this.

**Review by himanshu-garg-165 — CHANGES_REQUESTED** [review]

File: apps/user/serializers/user.go (hunk 3 above)
Comment (himanshu-garg-165) [reply apps/user/serializers/user.go]: Then we should use `emailOTP` .

**Review by himanshu-garg-165 — CHANGES_REQUESTED** [review]

File: constants/constants.go (hunk 16 above)
Comment (chahat-arora) [reply constants/constants.go]: done.

File: constants/constants.go (hunk 15 above)
Comment (chahat-arora) [reply constants/constants.go]: done.

File: apps/user/services/authentication.go (hunk 14 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: done.

File: apps/user/services/authentication.go (hunk 13 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: yes.

File: apps/user/services/authentication.go (hunk 12 above)
Comment (chahat-arora) [reply apps/user/services/authentication.go]: done.

**Review by himanshu-garg-165 — APPROVED** [review]

**Pull request merged**